src/array/cpu/spmm.h

```diff
@@ -57,9 +60,90 @@ void SpMMSumCsr(
         const DType *rhs_off =
             Op::use_rhs ? W + eid * rhs_dim + rhs_add : nullptr;
         out_off[k] += Op::Call(lhs_off, rhs_off);
+          }
+       }
+    }
+  }
+};
+
+/*!
+ * \brief CPU kernelCore of SpMM<float> specialization CopyLhs.
+ * \param bcast Broadcast information.
+ * \param csr The Csr matrix.
+ * \param ufeat The feature on source nodes.
+ * \param efeat The feature on edges.
+ * \param out The result feature on destination nodes.
+ * \note it uses node parallel strategy, different threads are responsible
+ *       for the computation of different nodes.
+ */
+#if !defined(_WIN32)
+template <class IdType>
+struct SpMMSumCsrCore<IdType, float, dgl::aten::cpu::op::CopyLhs<float>> {
+        static inline void call(const BcastOff &bcast,
+                                const CSRMatrix &csr,
+                                NDArray ufeat, NDArray efeat,
+                                NDArray out) {
+          typedef float DType;
+          const IdType *indptr = csr.indptr.Ptr<IdType>();
+          const IdType *indices = csr.indices.Ptr<IdType>();
+          const DType *X = ufeat.Ptr<DType>();
+          int64_t dim = bcast.out_len,
+                  lhs_dim = bcast.lhs_len;
+          DType *O = out.Ptr<DType>();
+          typedef intel::elem_wise_add_update<DType> ElemWiseUpd;
+
+          /* Prepare an assembler code for this paricular dimension */
+          thread_local intel::Uptr<ElemWiseUpd> th_cpu_spec((intel::IntelKernel<DType>::enabled())
+           ? new ElemWiseUpd(dim) : nullptr);
+          /* Check if the code has been created */
+          ElemWiseUpd *cpu_spec = (th_cpu_spec && th_cpu_spec->is_applicable())
+           ? th_cpu_spec.get() : nullptr;
+          if (cpu_spec && cpu_spec->require_new_instance<bool>(dim)) {
+            th_cpu_spec.reset(new ElemWiseUpd(dim));
+            cpu_spec = (th_cpu_spec && th_cpu_spec->is_applicable()) ? th_cpu_spec.get() : nullptr;
+          }
+
+#pragma omp parallel for
+  for (IdType rid = 0; rid < csr.num_rows; ++rid) {
+    const IdType row_start = indptr[rid], row_end = indptr[rid + 1];
+    DType *out_off = O + rid * dim;
+    std::fill(out_off, out_off + dim, 0);
+    for (IdType j = row_start; j < row_end; ++j) {
+      const IdType cid = indices[j];
+      if (!bcast.use_bcast && cpu_spec) {
+          /* Forward to cpu asm kernel */
+          cpu_spec->run(out_off, X + cid * lhs_dim , dim);
+          continue;
+      }
```

Comment:
Can we lift this branch out of the loop because it does not depend on the loop variable.

Reply:
done